source/slang/slang-ir-legalize-global-values.cpp

```diff
@@ -0,0 +1,261 @@
+#include "slang-ir-legalize-global-values.h"
+
+#include "slang-ir-clone.h"
+#include "slang-ir-util.h"
+
+namespace Slang
+{
+
+// Returns true iff 'rootInst', or any of it's descendants, satisfies 'predicate'.
+template<typename P>
+bool hasDescendantSatisfyingPredicate(IRInst* rootInst, P predicate)
+{
+    IRInst* inst = rootInst;
+    while (true)
+    {
+        if (predicate(inst))
+            return true;
+
+        // Step to next IRInst according to preorder
+        if (inst->getFirstDecorationOrChild())
+            inst = inst->getFirstDecorationOrChild();
+        else
+        {
+            while (true)
+            {
+                if (inst == rootInst)
+                    return false;
+                if (inst->getNextInst())
+                    break;
+                inst = inst->getParent();
+            }
+            inst = inst->getNextInst();
+        }
+    }
+    return false;
+}
+
+void GlobalInstInliningContextGeneric::inlineGlobalValuesAndRemoveIfUnused(IRModule* module)
+{
+    List<IRUse*> globalInstUsesToInline;
+    List<IRInst*> globalInstsToConsiderDeleting;
+
+    for (auto globalInst : module->getGlobalInsts())
+    {
+        if (isInlinableGlobalInst(globalInst))
+        {
+            for (auto use = globalInst->firstUse; use; use = use->nextUse)
+            {
+                if (getParentFunc(use->getUser()) != nullptr)
+                    globalInstUsesToInline.add(use);
+            }
+            globalInstsToConsiderDeleting.add(globalInst);
+        }
+    }
+
+    for (auto use : globalInstUsesToInline)
+    {
+        auto user = use->getUser();
+        IRBuilder builder(user);
+        builder.setInsertBefore(getOutsideASM(user));
+        IRCloneEnv cloneEnv;
+        auto val = maybeInlineGlobalValue(builder, use->getUser(), use->get(), cloneEnv);
+        if (val != use->get())
+            builder.replaceOperand(use, val);
+    }
+
+    // Since certain globals that appear in the IR are considered illegal for all targets,
+    // e.g. calls to functions, we delete globals which no longer have uses (or children with uses)
+    // after inlining.
+    for (auto& globalInst : globalInstsToConsiderDeleting)
+    {
+        if (!hasDescendantSatisfyingPredicate(
```

Comment:
This is a band aid instead of an actual fix.

As we discussed in the meeting, the true problem is when we clone the global SpirvASM inst to a function, we didn't update the operand references to point to the cloned version of the local spirv operand insts, as a result, the inlined asm block is still referencing the child insts in the original global asm block.

The fix here just prevents removal of the global asm block, but the right fix is to make sure we dont have such inter block references back to the original global asm block.

Reply:
Ok, I must have misunderstood something. 

Yesterday I stepped through the inlining code in the debugger and noticed that the `SPIRVAsmOperandInst`s didn't get inlined (unlike all of the other operands) and so there was no clone to be found.

I figured this is something that can happen in general, and so we must check for uses recursively before deleting the global inst.

So you are saying that the existing inlining logic is broken?

I will take another look.

Reply:
@csyonghe Should the fix be that, when we decide to inline SPIRV ASM insts then we force-inline things it contains, like the `SPIRVAsmOperandInst`s etc.. ?

Reply:
When we looked through the code together, it appears to me that the SPIRVAsm parent inst is being cloned, and its children are also being cloned. It is just the operands of the clone seems to be still referencing the original ones instead of the cloned ones.

e.g.

```
%b = SPIRV_Asm {
    %x = SPIRVAsmOperandInst ...
    %y = SPIRVAsmOperandInst ...
     %z = SPIRVAsmInst %x %y
}
```

got cloned into `main` as:

```
func %main  {
%block = block {
    %b' = SPIRV_Asm {
       %x' = SPIRVAsmOperandInst ...
       %y' = SPIRVAsmOperandInst ...
        %z' = SPIRVAsmInst %x %y // !!! should be %x' here, instead of %x
    }
}
}
```

Reply:
The reason that the SPIRV Operands are not inlined is that they are legal global insts, according to this:

https://github.com/shader-slang/slang/blob/772c48af91951bda9ff7064fc7a089550c759f76/source/slang/slang-ir-spirv-legalize.cpp#L1614

All of the other operands of the SPIRVAsmInst being inlined are not considered legal global insts, and so they are inlined.

Reply:
Oh, I think I see now!

The `SPIRVAsmOperand`s are special in that they are parented to the `SPIRVAsm` which we would like to delete.

The `inlineInst` function does indeed inline its children, but then later when we get to the last child which is a `SPIRVAsmInst` taking `SPIRVAsmOperand`s as operands, we decide not to inline them.

I see that there is a dictionary called `m_mapGlobalInstToShouldInline` that is first consulted.
Perhaps whenever we inline some inst we should set it as `true` in `m_mapGlobalInstToShouldInline`...

Reply:
> Oh, I think I see now!
> 
> The `SPIRVAsmOperand`s are special in that they are parented to the `SPIRVAsm` which we would like to delete.
> 
> The `inlineInst` function does indeed inline its children, but then later when we get to the last child which is a `SPIRVAsmInst` taking `SPIRVAsmOperand`s as operands, we decide not to inline them.
> 
> I see that there is a dictionary called `m_mapGlobalInstToShouldInline` that is first consulted. Perhaps whenever we inline some inst we should set it as `true` in `m_mapGlobalInstToShouldInline`...

That works, but only because the SPIRVAsmInst taking these SPIRVAsmOperandInst's as operands happen to appear after the SPIRVAsmOperandInst's children of the SPIRVAsm... 

If I instead mark `m_mapGlobalInstToShouldInline` to true for any inlined insts first thing in `inlineInst`, and then I can change `shouldInlineInstImpl` to return `true` in case any ancestors of the inst in question should be inlined according to `m_mapGlobalInstToShouldInline`.

Reply:
I did something similar to the above, and it seems to work.